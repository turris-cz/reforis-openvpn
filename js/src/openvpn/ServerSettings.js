/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useContext } from "react";
import PropTypes from "prop-types";

import {
    CheckBox, Button, Spinner, useForm, useAPIPatch, Select, TextInput,
    validateIPv4Address, AlertContext, undefinedIfEmpty, withoutUndefinedKeys, onlySpecifiedKeys,
    Input,
} from "foris";

import API_URLs from "API";

ServerSettings.propTypes = {
    settingsData: PropTypes.object.isRequired,
};

export default function ServerSettings({ settingsData }) {
    const setAlert = useContext(AlertContext);

    const [patchSettingsResponse, patchSettings] = useAPIPatch(API_URLs.serverSettings);
    useEffect(() => {
        // Happy path is omited since network restart causes the page to reload
        if (patchSettingsResponse.isError) {
            setAlert(_("Cannot save server settings"));
        }
    }, [patchSettingsResponse, setAlert]);

    const [formState, formChangeHandler] = useForm(validator);
    const formData = formState.data;
    const formErrors = formState.errors || {};
    useEffect(() => {
        const eventHandler = formChangeHandler((value) => ({ $set: { ...value } }));
        eventHandler({ target: { value: settingsData } });
    }, [settingsData, formChangeHandler]);

    function handleSubmit(event) {
        event.preventDefault();
        if (formData.enabled) {
            patchSettings(onlySpecifiedKeys(
                formData,
                ["enabled", "ipv6", "protocol", "network", "network_netmask", "route_all", "use_dns"],
            ));
        } else {
            patchSettings({ enabled: false });
        }
    }

    if (!formData) {
        return (
            <>
                <h3 className="mb-3">{_("Server settings")}</h3>
                <Spinner className="text-center" />
            </>
        );
    }

    return (
        <>
            <h3 className="mb-3">{_("Server settings")}</h3>
            <p>{_("Please note that you need a public (preferably static) IP address and your network configured to make use of VPN server.")}</p>
            <form onSubmit={handleSubmit} className="mb-3">
                <CheckBox
                    label={_("Server enabled")}
                    checked={formData.enabled}
                    onChange={formChangeHandler((value) => ({ enabled: { $set: value } }))}
                />
                {formData.enabled && (
                    <>
                        <Input type="text" label={_("Device")} value={formData.device} disabled />
                        <CheckBox
                            label={_("Listen on IPv6")}
                            checked={formData.ipv6}
                            onChange={formChangeHandler((value) => ({ ipv6: { $set: value } }))}
                            helpText={_("Useful if you don't have public IPv4 address. Disable if your IPv6 connection is broken.")}
                        />
                        <Select
                            label={_("Protocol")}
                            choices={{ udp: "UDP", tcp: "TCP" }}
                            value={formData.protocol}
                            onChange={formChangeHandler((value) => ({ protocol: { $set: value } }))}
                            helpText={_("Clients will connect to the VPN server via this protocol. Use TCP if your ISP blocks UDP.")}
                        />
                        <TextInput
                            label={_("VPN network address")}
                            value={formData.network}
                            error={formErrors.network}
                            onChange={formChangeHandler((value) => ({ network: { $set: value } }))}
                            helpText={_("This network should be different than any network directly reachable by the router and the clients.")}
                        />
                        <TextInput
                            label={_("VPN network mask")}
                            value={formData.network_netmask}
                            error={formErrors.network_netmask}
                            onChange={
                                formChangeHandler((value) => ({ network_netmask: { $set: value } }))
                            }
                        />
                        <Input type="text" label={_("Port")} value={formData.port} disabled />
                        <CheckBox
                            label={_("Route all traffic via VPN")}
                            checked={formData.route_all}
                            onChange={
                                formChangeHandler((value) => ({ route_all: { $set: value } }))
                            }
                            helpText={_("Clients will access the Internet via VPN")}
                        />
                        <CheckBox
                            label={_("Use DNS via VPN")}
                            checked={formData.use_dns}
                            onChange={formChangeHandler((value) => ({ use_dns: { $set: value } }))}
                            helpText={_("Clients will resolve addresses using router's DNS. Required when accessing local servers like my-nas.lan")}
                        />
                    </>
                )}
                <Button type="submit" forisFormSize disabled={undefinedIfEmpty(formErrors) || patchSettingsResponse.isSending}>Save</Button>
            </form>
            <p dangerouslySetInnerHTML={{ __html: _("<strong>Advanced users:</strong> if you already configured OpenVPN server manually your config will be extended (rather than  overwritten). In case of conflict you must fix previous settings by yourself.") }} />
        </>
    );
}

function validator(formData) {
    const errors = {
        network: validateIPv4Address(formData.network),
        network_netmask: validateIPv4Address(formData.network_netmask),
    };
    return undefinedIfEmpty(withoutUndefinedKeys(errors));
}
