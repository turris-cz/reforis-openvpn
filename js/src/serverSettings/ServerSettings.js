/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import { useAPIGet, withErrorMessage, withSpinnerOnSending } from "foris";

import API_URLs from "API";
import AuthorityStatus from "./AuthorityStatus";
import ServerSettingsForm from "./ServerSettingsForm";

ServerSettings.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function ServerSettings({ ws }) {
    const [authorityResponse, getAuthority] = useAPIGet(API_URLs.authority);
    useEffect(() => {
        getAuthority();
    }, [getAuthority]);

    const [settingsResponse, getSettings] = useAPIGet(API_URLs.serverSettings);
    useEffect(() => {
        getSettings();
    }, [getSettings]);

    return (
        <>
            <h1>{_("Server Settings")}</h1>
            <SettingsWithErrorAndSpinner
                apiState={[authorityResponse.state, settingsResponse.state]}
                ws={ws}
                authorityData={authorityResponse.data}
                settingsData={settingsResponse.data}
                onReload={getAuthority}
            />
        </>
    );
}

const SettingsWithErrorAndSpinner = withErrorMessage(withSpinnerOnSending(Settings));

Settings.propTypes = {
    ws: PropTypes.object.isRequired,
    authorityData: PropTypes.object.isRequired,
    settingsData: PropTypes.object.isRequired,
    onReload: PropTypes.func.isRequired,
};

function Settings({
    ws, authorityData, settingsData, onReload,
}) {
    return (
        <AuthorityStatus
            ws={ws}
            status={authorityData.status}
            serverEnabled={settingsData.enabled}
            onReload={onReload}
        >
            <ServerSettingsForm settingsData={settingsData} />
        </AuthorityStatus>
    );
}
