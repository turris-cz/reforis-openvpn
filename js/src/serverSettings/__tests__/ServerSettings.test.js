/*
 * Copyright (C) 2019-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    render,
    waitForElement,
    fireEvent,
    wait,
    getByText,
} from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from "jest-mock-axios";

import { WebSockets } from "foris";

import ServerSettings from "../ServerSettings";

describe("<ServerSettings />", () => {
    let componentContainer;

    beforeEach(() => {
        const webSockets = new WebSockets();
        const { container } = render(<ServerSettings ws={webSockets} />);
        componentContainer = container;
    });

    it("should display spinner", async () => {
        expect(componentContainer).toMatchSnapshot();
    });

    it("should handle API error", async () => {
        mockJSONError();
        await wait(() =>
            expect(
                getByText(
                    componentContainer,
                    "An error occurred while fetching data."
                )
            ).toBeDefined()
        );
    });

    it("should render child and set alert", async () => {
        expect(getByText(componentContainer, "Server Settings")).toBeDefined();

        expect(mockAxios.get).toBeCalledWith(
            "/reforis/openvpn/api/authority",
            expect.anything()
        );
        mockAxios.mockResponse({ data: { status: "ready" } });
        expect(mockAxios.get).toBeCalledWith(
            "/reforis/openvpn/api/server-settings",
            expect.anything()
        );
        mockAxios.mockResponse({ data: { enabled: false } });
        await waitForElement(() =>
            getByText(componentContainer, "Certificate Authority")
        );

        fireEvent.click(getByText(componentContainer, "Delete CA"));
        expect(mockAxios.delete).toBeCalledWith(
            "/reforis/openvpn/api/authority",
            expect.anything()
        );
        mockJSONError();
        await wait(() => {
            expect(mockSetAlert).toBeCalledWith(
                "Cannot delete certificate authority"
            );
        });
        expect(componentContainer).toMatchSnapshot();
    });
});
