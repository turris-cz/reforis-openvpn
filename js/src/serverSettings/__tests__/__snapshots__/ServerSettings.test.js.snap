// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ServerSettings /> should display spinner 1`] = `
<div>
  <h1>
    Server Settings
  </h1>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;

exports[`<ServerSettings /> should render child and set alert 1`] = `
<div>
  <h1>
    Server Settings
  </h1>
  <div
    class="card p-4 mb-3"
  >
    <h2
      class="mb-3"
    >
      Server settings
    </h2>
    <p>
      Please note that you need a public (preferably static) IP address and your network configured to make use of VPN server.
    </p>
    <form>
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-switch custom-control-inline"
        >
          <input
            class="custom-control-input"
            id="1"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="1"
          >
            Server enabled
          </label>
        </div>
      </div>
      <div
        class="text-right"
      >
        <button
          class="btn btn-primary  col-sm-12 col-md-3 col-lg-2"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
  <p>
    <strong>
      Advanced users:
    </strong>
     if you already configured OpenVPN server manually your configuration will be extended (rather than  overwritten). In case of conflict you must fix previous settings by yourself.
  </p>
  <div
    class="card p-4"
  >
    <h2>
      Certificate authority
    </h2>
    <p>
      Your certificate authority (CA) is set up properly. Please note that if you delete it all clients will have their access revoked.
    </p>
    <div
      class="text-right"
    >
      <button
        class="btn btn-danger col-sm-12 col-md-3 col-lg-2"
        type="button"
      >
        Delete CA
      </button>
    </div>
  </div>
</div>
`;
